src/Compilers/CSharp/Portable/Symbols/FunctionPointers/FunctionPointerMethodSymbol.cs

```diff
@@ -263,6 +263,70 @@ static void checkUnmanagedSupport(CSharpCompilation compilation, Location errorL
                 compilation);
         }
 
+        /// <summary>
+        /// Creates a function pointer method symbol from individual parts. This method should only be used when diagnostics are not needed.
+        /// </summary>
+        internal static FunctionPointerMethodSymbol CreateFromParts(
+            CallingConvention callingConvention,
+            ImmutableArray<CustomModifier> callingConventionModifiers,
+            TypeWithAnnotations returnTypeWithAnnotations,
+            RefKind returnRefKind,
+            ImmutableArray<TypeWithAnnotations> parameterTypes,
+            ImmutableArray<RefKind> parameterRefKinds,
+            CSharpCompilation compilation)
+        {
+            var modifiersBuilder = ArrayBuilder<CustomModifier>.GetInstance();
+
+            if (!callingConventionModifiers.IsDefaultOrEmpty)
+            {
+                Debug.Assert(callingConvention == CallingConvention.Unmanaged);
+                modifiersBuilder.AddRange(callingConventionModifiers);
+            }
+
+            ImmutableArray<CustomModifier> refCustomModifiers;
+            if (returnRefKind == RefKind.None)
+            {
+                refCustomModifiers = ImmutableArray<CustomModifier>.Empty;
+                returnTypeWithAnnotations = returnTypeWithAnnotations.WithModifiers(modifiersBuilder.ToImmutableAndFree());
+            }
+            else
+            {
+                if (GetCustomModifierForRefKind(returnRefKind, compilation) is CustomModifier modifier)
+                {
+                    modifiersBuilder.Add(modifier);
+                }
+                refCustomModifiers = modifiersBuilder.ToImmutableAndFree();
+            }
```

Comment:
>} [](start = 12, length = 1)

I'm confused. Why are `callingConventionModifiers` sometimes associated with the return type and sometimes associated with `refCustomModifiers`?

Reply:
The modifiers are always the first modifiers after the param count in metadata. If the refkind is `None`, then those modifiers are on the type. The refkind is not `None`, then those modifiers are on the ref specifier.

---
In reply to: [467328871](https://github.com/dotnet/roslyn/pull/46467#discussion_r467328871) [](ancestors = 467328871)